docs/book/src/reference/testdata/watching/controller.go

```diff
@@ -0,0 +1,218 @@
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch
+
+/*
+`Reconcile` will be in charge of reconinciling the state of ConfigDeployments.
+ConfigDeployments are used to manage Deployments whose pods are updated whenever the configMap that they use is updated.
+
+For that reason we need to add an annotation to the PodTemplate within the Deployment we create.
+This annotation will keep track of the latest version of the data within the referenced ConfigMap.
+Therefore when the version of the configMap is changed, the PodTemplate in the Deployment will change.
+This will cause a rolling upgrade of all Pods managed by the Deployment.
+
+Skip down to the `SetupWithManager` function to see how we ensure that `Reconcile` is called when the referenced `ConfigMaps` are updated.
+*/
+func (r *ConfigDeploymentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	/* */
+	_ = context.Background()
+	_ = r.Log.WithValues("configdeployment", req.NamespacedName)
+
+	var configDeployment appsv1.ConfigDeployment
+	if err := r.Get(ctx, req.NamespacedName, &configDeployment); err != nil {
+		log.Error(err, "unable to fetch ConfigDeployment")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	// +kubebuilder:docs-gen:collapse=Begin the Reconcile
+
+	// your logic here
+
+	var configMapVersion string
+	if configDeployment.Spec.ConfigMap != "" {
+		configMapName := configDeployment.Spec.ConfigMap
+		foundConfigMap := &corev1.ConfigMap{}
+		err := r.Get(context.TODO(), types.NamespacedName{Name: configMapName, Namespace: configDeployment.Namespace}, foundConfigMap)
+		if err != nil {
+			// If a configMap name is provided, then it must exist
+			// You will likely want to create an Event for the user to understand why their reconcile is failing.
+			return ctrl.Result{}, err
+		}
+
+		// Hash the data in some way, or just use the version of the Object
+		configMapVersion = foundConfigMap.ResourceVersion
+	}
+
+	// Logic here to add the configMapVersion as an annotation on your Deployment Pods.
+
+	/* */
+
+	return ctrl.Result{}, nil
+}
+// +kubebuilder:docs-gen:collapse=Finish the Reconcile
+
+/*
+Finally, we add this reconciler to the manager, so that it gets started
+when the manager is started.
+
+Since we create dependency Deployments during the reconcile, we can specify that the controller `Owns` Deployments.
+
+However the ConfigMaps that we want to watch are not owned by the ConfigDeployment object.
+Therefore we must specify a custom way of watching those objects.
+This watch logic is complex, so we have split it into a separate method.
```

Comment:
IMO it is a bit unclear extracting all to a function. I would:
- Keep the key as a `const` at the top of the file.
- Set the index first similar to how we do it in [the CronJob tutorial](https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#setup), before calling `ctr.NewControllerManagedBy`.
- Call the `.Watches` here and extract to a separate function only what you have as an anonymous function below (inside `handler.ToRequestsFunc`).

Reply:
I have refactored this method, so it should now be how you have described.

Let me know if it's still unclear.

Reply:
Perfect, you can resolve this conversation.